protocol/objects/block_header.ts

```diff
@@ -0,0 +1,132 @@
+import { AdDigest, adDigest, Digest32, digest32 } from "../../crypto/mod.ts";
+import { CursorReader, CursorWriter } from "../../io/cursor_buffer.ts";
+import {
+  NetworkObject,
+  OBJECT_ID_LENGTH,
+  ObjectId,
+  objectIdFromBytes,
+  objectIdToBytes,
+  ObjectTypeId,
+} from "./object.ts";
+
+export enum BlockVersion {
+  /** Block version during mainnet launch */
+  Initial = 1,
+  /** Block version after the Hardening hard-fork Autolykos v2 PoW, witnesses in transactions Merkle tree */
+  Hardening = 2,
+  /** Block version after the 5.0 soft-fork 5.0 interpreter with JITC, monotonic height rule (EIP-39) */
+  Interpreter50 = 3,
+}
+
+interface BlockHeaderOpts {
+  version: BlockVersion;
+  parentId: ObjectId;
+  adProofRoot: Digest32;
+  stateRoot: AdDigest;
+  txRoot: Digest32;
+  timestamp: bigint;
+  nBits: number;
+  height: number;
+  extensionRoot: Digest32;
+  // powSolution make optional or separate class like ref client?
+  /** Votes are 3 bytes in size */
+  votes: Uint8Array;
+}
+
+/**
+ * Block header network object sent and received over the Ergo network.
+ *
+ * NOTE: The reference client uses 2 separate types, a header with pow solution
+ * and a header without pow solution, not sure why this is needed at the moment
+ * so just going with the one type for now.
+ */
+export class BlockHeader implements NetworkObject {
```

Comment:
Move everything in here to a new `chain` module and create dedicated serialize/deserialize classes for sending chain objects over the wire.

Reply:
Then in here we would have a `BlockHeaderNetworkObject` that wraps the `BlockHeader` type and performs encoding/decoding & returns the `objectTypeId`